velox/functions/prestosql/InPredicate.cpp

```diff
@@ -19,21 +19,124 @@
 namespace facebook::velox::functions {
 namespace {
 
+// Returns NULL if
+// - input value is NULL or contains NULL;
+// - input value doesn't match any of the in-list values, but some of in-list
+// values are NULL or contain NULL.
+class ComplexTypeInPredicate : public exec::VectorFunction {
+ public:
+  struct ComplexValue {
+    BaseVector* vector;
+    vector_size_t index;
+  };
+
+  struct ComplexValueHash {
+    size_t operator()(ComplexValue value) const {
+      return value.vector->hashValueAt(value.index);
+    }
+  };
+
+  struct ComplexValueEqualTo {
+    bool operator()(ComplexValue left, ComplexValue right) const {
+      return left.vector->equalValueAt(right.vector, left.index, right.index);
+    }
+  };
+
+  using ComplexSet =
+      folly::F14FastSet<ComplexValue, ComplexValueHash, ComplexValueEqualTo>;
+
+  ComplexTypeInPredicate(
+      ComplexSet uniqueValues,
+      bool hasNull,
+      VectorPtr originalValues)
+      : uniqueValues_{std::move(uniqueValues)},
+        hasNull_{hasNull},
+        originalValues_{std::move(originalValues)} {}
+
+  static std::shared_ptr<ComplexTypeInPredicate> create(
+      const VectorPtr& values) {
+    ComplexSet uniqueValues;
+    bool hasNull = false;
+
+    VELOX_CHECK_EQ(values->typeKind(), TypeKind::ARRAY);
+
+    auto constantInput =
+        std::dynamic_pointer_cast<ConstantVector<ComplexType>>(values);
```

Comment:
n00b question, but does IN() only support constant literals? What guarantees this vector is always constant-encoded?

Reply:
probably worth asserting this pointer is not null, in any case, and throwing a descriptive message. If we guarantee in the caller this never happens, then maybe we could change the API to take a constant vector instead.

Reply:
just saw below that this is the latter case.

Reply:
> What guarantees this vector is always constant-encoded?

The signature is defined using .constantArgumentType("array(T)").

> but does IN() only support constant literals? 

It should support non-constant inputs as well. We'll need to add that logic.